src/Rings/msolve/f4.jl

```diff
@@ -0,0 +1,121 @@
+See [Fau99](@cite) for more information.
+
+**Note**: At the moment only ground fields of characteristic `p`, `p` prime, `p < 2^{31}` are supported.
+
+# Arguments
+- `Ì::MPolyIdeal`: input ideal.
+- `initial_hts::Int=17`: initial hash table size `log_2`.
+- `nr_thrds::Int=1`: number of threads for parallel linear algebra.
+- `max_nr_pairs::Int=0`: maximal number of pairs per matrix, only bounded by minimal degree if `0`.
+- `la_option::Int=2`: linear algebra option: exact sparse-dense (`1`), exact sparse (`2`, default), probabilistic sparse-dense (`42`), probabilistic sparse(`44`).
+- `reduce_gb::Int=1`: compute a reduced Gröbner basis for `I`
+- `info_level::Int=0`: info level printout: off (`0`, default), summary (`1`), detailed (`2`).
+
+# Examples
+```jldoctest
+julia> R,(x,y,z) = PolynomialRing(FiniteField(101), ["x","y","z"])
+(Multivariate Polynomial Ring in x, y, z over Galois field with characteristic 101, gfp_mpoly[x, y, z])
+
+julia> I = ideal(R, [x+2*y+2*z-1, x^2+2*y^2+2*z^2-x, 2*x*y+2*y*z-y])
+ideal(x + 2*y + 2*z + 100, x^2 + 100*x + 2*y^2 + 2*z^2, 2*x*y + 2*y*z + 100*y)
+
+julia> f4(I)
+Oscar.BiPolyArray{gfp_mpoly}(gfp_mpoly[#undef, #undef, #undef, #undef], Singular ideal over Singular Polynomial Ring (ZZ/101),(x,y,z),(dp(3),C) with generators (x + 2*y + 2*z - 1, y*z - 19*z^2 + 10*y + 40*z, y^2 - 41*z^2 + 20*y - 20*z, z^3 + 28*z^2 - 37*y + 13*z), Multivariate Polynomial Ring in x, y, z over Galois field with characteristic 101, Singular Polynomial Ring (ZZ/101),(x,y,z),(dp(3),C), true, #undef)
+```
+"""
+function f4(
+        I::MPolyIdeal;
+        initial_hts::Int=17,
+        nr_thrds::Int=1,
+        max_nr_pairs::Int=0,
+        la_option::Int=2,
+        reduce_gb::Int=1,
+        info_level::Int=0
+        )
+
+    singular_assure(I)
+    SI    = I.gens.S
+    R     = I.gens.Sx
+    # skip zero generators in ideal
+    ptr = Singular.libSingular.id_Copy(SI.ptr, R.ptr)
+    J   = Singular.Ideal(R, ptr)
+    Singular.libSingular.idSkipZeroes(J.ptr)
+    # get number of variables
+    nr_vars = Singular.nvars(R)
+    nr_gens = Singular.ngens(J)
+
+    mon_order   = 0
+    field_char  = Singular.characteristic(R)
+
+    # convert Singular ideal to flattened arrays of ints
+    if isprime(field_char)
+      lens, cfs, exps = convert_ff_singular_ideal_to_array(J, nr_vars, nr_gens)
+    else
+        error("At the moment f4 only supports finite fields.")
+    end
+    #= dir = joinpath(dirname(pathof(GroebnerBasis)),"../deps")
+     = lib = Libdl.dlopen("$dir/libmsolve.so.0.2.0") =#
+    lib = Libdl.dlopen(libneogb)
+    sym = Libdl.dlsym(lib, :f4_julia)
+
+    gb_ld   = ccall(:malloc, Ptr{Cint}, (Csize_t, ), sizeof(Cint))
+    gb_len  = ccall(:malloc, Ptr{Ptr{Cint}}, (Csize_t, ), sizeof(Ptr{Cint}))
+    gb_exp  = ccall(:malloc, Ptr{Ptr{Cint}}, (Csize_t, ), sizeof(Ptr{Cint}))
+    gb_cf   = ccall(:malloc, Ptr{Ptr{Cvoid}}, (Csize_t, ), sizeof(Ptr{Cvoid}))
+    nr_terms  = ccall(sym, Int,
```

Comment:
create a c-struct to transport the variables? might be slightly more readable...